src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorFn.java

```
@@ -81,167 +59,118 @@
-
-        baq = new BAQ(BRAC.BAQGOP); // setup the BAQ object with the provided gap open penalty
-
-        if (BRAC.RAC.FORCE_PLATFORM != null)
-            BRAC.RAC.DEFAULT_PLATFORM = BRAC.RAC.FORCE_PLATFORM;
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
 
-        covariates = new StandardCovariateList(BRAC.RAC, getHeaderForReads());
-
-        initializeRecalibrationEngine();
-        minimumQToUse = BRAC.PRESERVE_QSCORES_LESS_THAN;
-
-        if (null!=refFile) {
-            referenceDataSource = new ReferenceDataSource(refFile);
+        // can't set things up in startBundle because it doesn't have our side input yet.
+        // So instead we're doing it here in processElement.
+        if (firstInBundle) {
+            init(c);
+            onTraversalStart();
+            firstInBundle = false;
         }
+        GATKRead r = c.element().getKey();
+        ReadContextData rc = c.element().getValue();
+        Iterable<Variant> variants = rc.getOverlappingVariants();
+        final ReferenceBases refBases = rc.getOverlappingReferenceBases();
+        ReferenceDataSource refDS = new ReferenceDataSourceFromReferenceBases(refBases, referenceSequenceDictionary);
+        ReferenceContext ref = new ReferenceContext(refDS, refBases.getInterval());
+
+        apply(r, ref, refDS, variants);
+    }
+
+    @Override
+    public void finishBundle(Context c) throws Exception {
+        bunny.stepEnd("finishBundle");
+        if (null!=recalibrationEngine) {
+            final RecalibrationTables recalibrationTables = recalibrationEngine.getRecalibrationTables();
+            if (null==recalibrationTables) {
+                bunny.stepEnd("null output(!)");
+            }
+            c.output(recalibrationTables);
+        } else {
+            // this can happen, but only if we processed zero read in this bundle
+            if (nReadsProcessed!=0) {
+                System.err.println("Oops, something fishy is going on");
```

Throw an exception here rather than printing to `System.err`